eclipse-theia/theia
# Dev-Container support

#### What it does

This implements a first version of the [Dev-Container Support](https://code.visualstudio.com/docs/devcontainers/containers) vscode offers for Theia. This can later also be used to forward arbritrary ports at container runtime

The [devcontainer.json format](https://containers.dev/implementors/json_reference/) offers alot of configuration option of which currently only a certain few are implemented.

Implemented properties are:
- `image`
- `dockerfile`/`build.dockerfile` (dockerfile support still needs some more testing and work regarding its options)
  - including `build.args` which are just passed to docker as is
- `forwardPorts`
- `mounts`

The communication with the theia backend inside the container is done through a docker exec instance to the ´dev-conatiner-server.ts´ which forwads all communcation to stdio.

Support for having multiple devcontainer files is also added

I would love some feedback on this, especially what missing features would be most important to make it most useful.

#### How to test

You need docker installed and started on your machine

1. open a folder in theia
2. create a new `.devcontainer` folder at the root
3. add a `devcontainer.json` and neccessary resources for it
here is an example file using a image wich is automaticly pulled by docker
```
{
    "name": "test_container",
    "image": "nikolaik/python-nodejs:python3.11-nodejs20-slim",
    "forwardPorts": [4000, "3000:3003"],
    "mounts": [{"source": "some/path", "target": "/workspaces", "type": "bind" }]

}

```
or here is a dockerfile with its respective devcontainer.json file for testing dockerfile support:
[devcontainer.json](https://github.com/eclipse-theia/theia/files/14236525/devcontainer.json)
[Dockerfile.txt](https://github.com/eclipse-theia/theia/files/14236526/Dockerfile.txt)

4. execute `Dev Containers: Reopen in Container` 
5. your workspace should then open inside the container



#### Follow-ups

Things not implemented yet which vscodes supports:
- Port forwarding at runtime (this feature is in general still missing in the remote feature) (its mostly done though, there will be a PR as soon as this one is merged)
- attaching to running containers
- Docker Compose support
- variables inside `dockerconfig.json`
- Showing the docker ouput, especially for when images are build through dockerfiles or pulled
- lots of smaller configuration options


#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
